crates/metadata/src/lib.rs

```diff
@@ -643,6 +652,20 @@ impl ContractBuilder {
     }
 }
 
+/// Checks whether the contract's ink! version is compatible with the cargo-contract
+/// binary.
+pub fn check_contract_ink_compatibility(ink_version: &Version) -> Result<()> {
+    let ink_minimal_version =
+        Version::parse("4.0.0-alpha.1").expect("Parsing version failed");
```

Comment:
here dependency is hardcoded,
optionally we could have compatibility.json file:
{
   "versions": [
        {
            "cargo-contract": ">=1.0.0,<2.0.0",
            "ink": ">=3.0.0,<4.0.0"
        },
        {
            "cargo-contract": ">=4.0.0",
            "ink": ">=4.0.0"
        }
    ]
}
 and include it to code (include_str)